faunadb-java/src/main/java/com/faunadb/client/types/Value.java

```diff
@@ -3,344 +3,207 @@
  *
- * <p>The result tree can be accessed using:</p>
- *
- * <pre>
- *   node.get("ref").asRef(); // {@link Ref}("some/ref")
- *   node.get("data").get("someKey").asString() // "string1"
- * </pre>
+ *   node.get(ref); // Ref("some/ref")
+ *   node.get(someKey); // "string1"
+ *   node.getOptional(nonExistingKey) // Optional.absent()
+ * }</pre>
+ * <p>
+ * The interface also has helpers to transverse values without {@link Field} references:
+ * <pre>{@code
+ *   node.at("ref").as(Codec.REF).get(); // Ref("some/ref")
+ *   node.at("data", "someKey").as(Codec.STRING).get() // "string1"
+ *   node.at("non-existing-key").as(Codec.LONG).getOptional() // Optional.absent()
+ * }</pre>
  *
- * <p><i>Reference</i>: <a href="https://faunadb.com/documentation/queries#values">FaunaDB Value Types</a></p>
+ * @see <a href="https://faunadb.com/documentation/queries#values">FaunaDB Value Types</a>
+ * @see Field
+ * @see Codec
  */
-public interface Value {
-  /**
-   * Coerces this node into a {@link String}.
-   * @return the string value of this node, or null.
-   */
-  String asString();
-
-  /**
-   * Coerces this node into a {@link Boolean}.
-   * @return the boolean value of this node, or null.
-   */
-  Boolean asBoolean();
-
-  /**
-   * Coerces this node into a {@link Long}.
-   * @return the long value of this node, or null.
-   */
-  Long asLong();
+@JsonDeserialize(using = Deserializer.ValueDeserializer.class)
+public abstract class Value extends Expr {
 
   /**
-   * Coerces this node into a {@link Double}.
-   * @return the double value of this node, or null.
-   */
-  Double asDouble();
-
-  /**
-   * Coerces this node into a {@link Instant}, if it is a {@link TsV}.
-   * @return the instant value of this node, or null.
-   */
-  Instant asTs();
-
-  /**
-   * Coerces this node into a {@link LocalDate}, if it is a {@link DateV}.
-   * @return the date value of this node, or null.
-   */
-  LocalDate asDate();
-
-  /**
-   * Coerces this node into an ordered list of nodes.
-   * @return an ordered list of response nodes, or null.
-   */
-  ImmutableList<Value> asArray();
-
-  /**
-   * Coerces this node into a dictionary of nodes.
-   * @return a dictionary of nodes, or null.
-   */
-  ImmutableMap<String, Value> asObject();
-
-  /**
-   * Coerces this node into a {@link Ref}.
-   * @return a Ref, or null.
-   */
-  Ref asRef();
-
-  /**
-   * Coerces this node into a {@link Set}.
-   * @return a Set, or null.
-   */
-  Set asSet();
-
-  /**
-   * Coerces this node into a {@link Page}.
-   * @return a Page, or null.
-   */
-  Page asPage();
-
-  /**
-   * Coerces this node into an {@link Instance}.
-   * @return an Instance, or null.
+   * Attempts to coerce this value using the {@link Codec} passed
+   *
+   * @param codec codec function to attempt coercion
+   * @return the {@link Result} of the coercion
+   * @see Codec
    */
-  Instance asInstance();
+  public final <T> Result<T> as(Codec<T> codec) {
```

Comment:
`to` is static and `as` is not on Field. They are different methods and can have the same name because they have the same signature (both receive a `Codec`).

`as` is used on pre-configured fields:
```java
private static final Field<Ref> REF_FIELD = Field.at("ref").as(REF);
```

`to` is used when you've transversing part of the value and you need to convert its last root:
```java
private static final Field<Value> DATA = Field.at("data");
private static final Field<ImmutableList<Ref>> REF_LIST = DATA.collect(Field.to(REF));

// or

Value taken = client.query(Take(Value(2), Arr(Value(1), Value(2), Value(3)))).get();
assertThat(taken.collect(Field.to(LONG)), contains(1L, 2L));
```

I can remove the `to` method, but that will require the user to always have a pre-defined field to collect data from arrays. Or, perhaps, I can make the `root` method public. Then syntax would look like this:

```
private static final Field<ImmutableList<Ref>> REF_LIST = DATA.collect(Field.root().as(REF));
// or
assertThat(taken.collect(Field.root().as(LONG)), contains(1L, 2L));
```

What option do you think is the best?